File: PADTiltViewControllerExample/PADTiltViewControllerExample/PADTiltTableViewControllerExample.m
//
//  PADTiltTableViewControllerExample.m
//  PADTiltViewControllerExample
//
//  Created by Pavel Dusatko on 5/3/14.
//  Copyright (c) 2014 Pavel Dusatko. All rights reserved.
//

#import "PADTiltTableViewControllerExample.h"
#import <PADTiltViewController/UIViewController+PADTiltAdditions.h>
#import "PADAppDelegate.h"

@interface PADTiltTableViewControllerExample () <UIGestureRecognizerDelegate>

@property (nonatomic, weak) IBOutlet UILabel *titleLabel;
@property (nonatomic, weak) IBOutlet UILabel *subtitleLabel;

@end

@implementation PADTiltTableViewControllerExample

#pragma mark - NSObject - Creating, Copying, and Deallocating Objects

- (void)dealloc {
    @try {
        [self removeObserver:self forKeyPath:NSStringFromSelector(@selector(pad_state)) context:NULL];
    }
    @catch (NSException *exception) {}
}

#pragma mark - UIViewController - Managing the View

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.pad_motionManager = [(PADAppDelegate *)[[UIApplication sharedApplication] delegate] motionManager];
    
    [self addObserver:self forKeyPath:NSStringFromSelector(@selector(pad_state)) options:NSKeyValueObservingOptionNew context:NULL];
    
    [self enableLongPressGestureRecognizer];
}

#pragma mark - Gesture recognizers

- (void)removeGestureRecognizers {
    for (UIGestureRecognizer *gestureRecognizer in self.view.gestureRecognizers) {
        if ([gestureRecognizer isKindOfClass:[UILongPressGestureRecognizer class]] || [gestureRecognizer isKindOfClass:[UITapGestureRecognizer class]] || [gestureRecognizer isKindOfClass:[UISwipeGestureRecognizer class]]) {
            [self.view removeGestureRecognizer:gestureRecognizer];
        }
    }
}

- (void)enableLongPressGestureRecognizer {
    UILongPressGestureRecognizer *longPressGestureRecognizer = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(startReceivingTiltUpdates:)];
    [self.view addGestureRecognizer:longPressGestureRecognizer];
}

- (void)enableLongPressTapAndSwipeGestureRecognizer {
    UILongPressGestureRecognizer *longPressGestureRecognizer = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(stopReceivingTiltUpdates:)];
    [self.view addGestureRecognizer:longPressGestureRecognizer];
    
    UITapGestureRecognizer *tapGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(stopReceivingTiltUpdates:)];
    [self.view addGestureRecognizer:tapGestureRecognizer];
    
    UISwipeGestureRecognizer *swipeGestureRecognizer = [[UISwipeGestureRecognizer alloc] initWithTarget:self action:@selector(stopReceivingTiltUpdates:)];
    swipeGestureRecognizer.delegate = self;
    swipeGestureRecognizer.direction = UISwipeGestureRecognizerDirectionUp | UISwipeGestureRecognizerDirectionDown;
    [self.view addGestureRecognizer:swipeGestureRecognizer];
}

#pragma mark - UIGestureRecognizerDelegate

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer {
    return YES;
}

#pragma mark - Target-Action

- (void)startReceivingTiltUpdates {
    [self pad_startReceivingTiltUpdates];
    
    [self removeGestureRecognizers];
    [self enableLongPressTapAndSwipeGestureRecognizer];
}

- (void)stopReceivingTiltUpdates {
    [self pad_stopReceivingTiltUpdates];
    
    [self removeGestureRecognizers];
    [self enableLongPressGestureRecognizer];
}

- (void)startReceivingTiltUpdates:(UIGestureRecognizer *)gestureRecognizer {
    if (gestureRecognizer.state == UIGestureRecognizerStateBegan) {
        [self startReceivingTiltUpdates];
    }
}

- (void)stopReceivingTiltUpdates:(UIGestureRecognizer *)gestureRecognizer {
    if (gestureRecognizer.state == UIGestureRecognizerStateBegan || gestureRecognizer.state == UIGestureRecognizerStateEnded) {
        [self stopReceivingTiltUpdates];
    }
}

#pragma mark - KVO

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    if ([keyPath isEqualToString:NSStringFromSelector(@selector(pad_state))]) {
        [self updateUserInterfaceForState:[object pad_state]];
    }
}

#pragma mark - Private

- (void)updateUserInterfaceForState:(NSInteger)state {
    if (state == PADTiltViewControllerStateInactive) {
        [[UIApplication sharedApplication] setStatusBarStyle:UIStatusBarStyleDefault];
        
        self.navigationController.navigationBar.barTintColor = nil;
        self.titleLabel.text = NSLocalizedString(@"Sensors Off", nil);
        self.titleLabel.textColor = [UIColor blackColor];
        self.subtitleLabel.text = NSLocalizedString(@"Long press the content to enable sensors", nil);
        self.subtitleLabel.textColor = [UIColor blackColor];
        
        [UIView animateWithDuration:0.35 animations:^{
            self.tabBarController.tabBar.alpha = 1.0;
        }];
    }
    else if (state == PADTiltViewControllerStateInitializing || state == PADTiltViewControllerStateCalibrating) {
        [[UIApplication sharedApplication] setStatusBarStyle:UIStatusBarStyleLightContent];
        
        self.navigationController.navigationBar.barTintColor = [UIColor lightGrayColor];
        self.titleLabel.text = NSLocalizedString(@"Calibrating Sensors", nil);
        self.titleLabel.textColor = [UIColor whiteColor];
        self.subtitleLabel.text = NSLocalizedString(@"Hold still for a moment", nil);
        self.subtitleLabel.textColor = [UIColor whiteColor];
        
        [UIView animateWithDuration:0.35 animations:^{
            self.tabBarController.tabBar.alpha = 0.0;
        }];
    }
    else if (state == PADTiltViewControllerStateActive) {
        [[UIApplication sharedApplication] setStatusBarStyle:UIStatusBarStyleLightContent];
        
        self.navigationController.navigationBar.barTintColor = [UIColor lightGrayColor];
        self.titleLabel.text = NSLocalizedString(@"Sensors On", nil);
        self.titleLabel.textColor = [UIColor whiteColor];
        self.subtitleLabel.text = NSLocalizedString(@"Tap or swipe the content to disable sensors", nil);
        self.subtitleLabel.textColor = [UIColor whiteColor];
    }
}

@end

File: Swift-version/Inpiration/PADTiltTableViewControllerExample.swift
//
//  PADTiltTableViewControllerExample.swift
//  Inpiration
//
//  Created by Alexander Hüllmandel on 10/31/14.
//  Copyright (c) 2014 Alexander Hüllmandel. All rights reserved.
//

import Foundation
import UIKit

class PADTiltTableViewControllerExample: PADTiltTableViewController, UIGestureRecognizerDelegate {
    
    @IBOutlet weak var titleLabel: UILabel!
    @IBOutlet weak var subtitleLabel: UILabel!
    
    //MARK: - NSObject - Creating, Copying, and Deallocating Objects
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self, name: kPadStateChanged, object: nil)
    }
    
    //MARK: - UIViewController - Managing the View
    override func viewDidLoad() {
        super.viewDidLoad()
        pad_motionManager = (UIApplication.sharedApplication().delegate as? AppDelegate)?.motionManager
        
        NSNotificationCenter.defaultCenter().addObserverForName(kPadStateChanged, object: nil, queue: NSOperationQueue.mainQueue()) { notification in
            let userInfo = notification.userInfo as Dictionary<String,Int!>
            if let padStateValue = userInfo[kNewPadState] {
                let newState = PADTiltViewControllerState(rawValue:padStateValue)
                self.updateUserInterfaceForState(newState!)
            }
        }
        
        enableLongPressGestureRecognizer()
    }
    
    //MARK: - Gesture recognizers
    func removeGestureRecognizers() {
        for gestureRecognizer in self.view.gestureRecognizers as [UIGestureRecognizer] {
            if (gestureRecognizer is UILongPressGestureRecognizer || gestureRecognizer is UITapGestureRecognizer
                || gestureRecognizer is UISwipeGestureRecognizer) {
                    self.view.removeGestureRecognizer(gestureRecognizer)
            }
        }
    }
    
    func enableLongPressGestureRecognizer() {
        let longPressGestureRecognizer = UILongPressGestureRecognizer(target: self, action: Selector("startReceivingTiltUpdates:"))
        self.view.addGestureRecognizer(longPressGestureRecognizer)
    }
    
    func enableLongPressTapAndSwipeGestureRecognizer() {
        let longPressGestureRecognizer = UILongPressGestureRecognizer(target: self, action: Selector("stopReceivingTiltUpdates:"))
        self.view.addGestureRecognizer(longPressGestureRecognizer)
        
        let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: Selector("stopReceivingTiltUpdates:"))
        self.view.addGestureRecognizer(tapGestureRecognizer)
        
        let swipeGestureRecognizer = UISwipeGestureRecognizer(target: self, action: Selector("stopReceivingTiltUpdates:"))
        swipeGestureRecognizer.delegate = self
        swipeGestureRecognizer.direction = .Up | .Down
        self.view.addGestureRecognizer(swipeGestureRecognizer)
    }
    
    //MARK: - UIGestureRecognizerDelegate
    func gestureRecognizer(gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWithGestureRecognizer otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        return true
    }
    
    //MARK: - Target-Action
    
    func startReceivingTiltUpdates() {
        pad_startReceivingTiltUpdates()
        
        removeGestureRecognizers()
        enableLongPressTapAndSwipeGestureRecognizer()
    }
    
    func stopReceivingTiltUpdates() {
        pad_stopReceivingTiltUpdates()
        
        removeGestureRecognizers()
        enableLongPressGestureRecognizer()
    }
    
    func startReceivingTiltUpdates(gestureRecognizer: UIGestureRecognizer) {
        if (gestureRecognizer.state == .Began) {
            startReceivingTiltUpdates()
        }
    }
    
    func stopReceivingTiltUpdates(gestureRecognizer: UIGestureRecognizer) {
        if (gestureRecognizer.state == .Began || gestureRecognizer.state == .Ended) {
            stopReceivingTiltUpdates()
        }
    }
    
    //MARK: - Private
    
    func updateUserInterfaceForState(state: PADTiltViewControllerState) {
        if (state == .Inactive) {
            UIApplication.sharedApplication().setStatusBarStyle(.Default, animated: false)
            
            if let navigationController = navigationController {
                navigationController.navigationBar.barTintColor = nil
            }
            titleLabel.text = NSLocalizedString("Sensors Off", comment:"")
            titleLabel.textColor = UIColor.blackColor()
            subtitleLabel.text = NSLocalizedString("Long press the content to enable sensors", comment:"")
            subtitleLabel.textColor = UIColor.blackColor()
            
            UIView.animateWithDuration(0.35) {
                self.tabBarController!.tabBar.alpha = 1.0
            }
        }
        else if (state == .Initializing || state == .Calibrating) {
            UIApplication.sharedApplication().setStatusBarStyle(.LightContent, animated: false)
            
            if let navigationController = navigationController {
                navigationController.navigationBar.barTintColor = UIColor.lightGrayColor()
            }
            titleLabel.text = NSLocalizedString("Calibrating Sensors", comment:"")
            titleLabel.textColor = UIColor.whiteColor()
            subtitleLabel.text = NSLocalizedString("Hold still for a moment", comment:"")
            subtitleLabel.textColor = UIColor.whiteColor()
            
            UIView.animateWithDuration(0.35) {
                self.tabBarController!.tabBar.alpha = 0.0
            }
        }
        else if (state == .Active) {
            UIApplication.sharedApplication().setStatusBarStyle(.LightContent, animated: false)
            
            if let navigationController = navigationController {
                navigationController.navigationBar.barTintColor = UIColor.lightGrayColor()
            }
            titleLabel.text = NSLocalizedString("Sensors On", comment:"")
            titleLabel.textColor = UIColor.whiteColor()
            subtitleLabel.text = NSLocalizedString("Tap or swipe the content to disable sensors", comment:"")
            subtitleLabel.textColor = UIColor.whiteColor()
        }
    }
}